src/litserve/server.py

```diff
@@ -122,41 +121,58 @@ def run_single_loop(lit_api: LitAPI, lit_spec: LitSpec, request_queue: Queue, re
                 lit_api.encode_response,
                 y,
             )
-
-            with contextlib.suppress(BrokenPipeError):
-                pipe_s.send((y_enc, LitAPIStatus.OK))
+            response_queue.put((uid, (y_enc, LitAPIStatus.OK)))
         except Exception as e:
             logger.exception(
                 "LitAPI ran into an error while processing the request uid=%s.\n"
                 "Please check the error trace for more details.",
                 uid,
             )
-            with contextlib.suppress(BrokenPipeError):
-                pipe_s.send((pickle.dumps(e), LitAPIStatus.ERROR))
+            err_pkl = pickle.dumps(e)
+            response_queue.put((uid, (err_pkl, LitAPIStatus.ERROR)))
 
 
 def run_batched_loop(
     lit_api: LitAPI,
     lit_spec: LitSpec,
     request_queue: Queue,
-    request_buffer: Dict,
+    response_queue: Queue,
     max_batch_size: int,
     batch_timeout: float,
 ):
     while True:
         batches = collate_requests(
             lit_api,
             request_queue,
-            request_buffer,
             max_batch_size,
             batch_timeout,
         )
         if not batches:
             continue
 
         logger.debug(f"{len(batches)} batched requests received")
-        inputs, pipes = zip(*batches)
+        uids, timestamps, inputs = zip(*batches)
+
+        # Iterate in reverse order to safely remove items while iterating
```

Comment:
I would more simply make a copy of timestamps (they are not a lot), build a list of indices that need to be evicted, and then create a new list of uids, inputs, and timestamps accordingly

Reply:
But more in general, this is not the correct time to do this.
We need to do this in collate_requests, because batches should not contain timed out requests anyway.

At this point collate_requests needs to return
batches, timed_out_uids
instead of just batches, so this function can return the HTTP exception.